## Deriving Data Solutions as per Customer Requirements
### The solutions are based on what questions an average customer can ask. Some of them are given below:
1) Let's say a client wants to install its carrier and they have given the specific latitude, longitude. They can ask for locations near those given coordinates <br>
2) They can ask for tower type they should go for to get maximum results

In [2]:
import pandas as pd
import numpy as np

In [3]:
df = pd.read_csv("D:/INTERNSHIP/db_final.csv")
df.head(10)

Town   Latitude  Longitude               Tower Type  Height  \
0  North Canaan  42.014722 -73.326389  self-supporting lattice   195.0   
1     Montville  41.468778 -72.203250  self-supporting lattice   300.0   
2       Clinton  41.275206 -72.497711  self-supporting lattice   195.0   
3    Middlebury  41.513611 -73.124167  self-supporting lattice   160.0   
4       Clinton  41.278814 -72.525975  self-supporting lattice    67.0   
5        Hamden  41.415569 -72.928889  self-supporting lattice   100.0   
6     Montville  41.473958 -72.115856  self-supporting lattice   260.0   
7   East Haddam  41.461336 -72.437750  self-supporting lattice   120.0   
8     Fairfield  41.210000 -73.261667  self-supporting lattice    84.0   
9       Meriden  41.557231 -72.789744  self-supporting lattice   120.0   

             carrier_1        carrier_2        carrier_3        carrier_4  \
0      Cingular @ 143'  T-Mobile @ 125'              NaN              NaN   
1      t-mobile @ 180'              NaN              NaN              NaN   
2       Verizon @ 162'    Sprint @ 182'  Cingular @ 152'              NaN   
3      Cingular @ 135'    dps @ 57-160'     Sprint @ 97'  T-Mobile @ 125'   
4       T-Mobile @ 60'    Verizon @ 50'              NaN              NaN   
5                  NaN              NaN              NaN              NaN   
6           dep @ 260'       dps @ 254'  T-Mobile @ 235'              NaN   
7              DESPP @              NaN              NaN              NaN   
8  Cingular/AT&T @ 84'    Sprint @ 101'   T-Mobile @ 95'              NaN   
9     clearwire @ 110'              NaN              NaN              NaN   

       carrier_5 carrier_6 carrier_7  New Georeferenced Column  Carriers  
0            NaN       NaN       NaN                         6         2  
1            NaN       NaN       NaN                         8         1  
2            NaN       NaN       NaN                         3         3  
3  Verizon @ 75'       NaN       NaN                         4         5  
4            NaN       NaN       NaN                         3         2  
5            NaN       NaN       NaN                         7         0  
6            NaN       NaN       NaN                         8         3  
7            NaN       NaN       NaN                         3         1  
8            NaN       NaN       NaN                         2         3  
9            NaN       NaN       NaN                         7         1

### The statistics for all the types of towers are given below

In [4]:
types = ["self-supporting lattice", "guyed lattice", "utility pole", "monopole", "power mount", "rooftop"]

for tower_type in types:
    print(tower_type)
    print(df[df["Tower Type"]==tower_type]["Carriers"].describe())
    print(df[df["Tower Type"]==tower_type]["Height"].describe())

self-supporting lattice
count    195.000000
mean       2.846154
std        1.895987
min        0.000000
25%        1.000000
50%        3.000000
75%        4.000000
max        7.000000
Name: Carriers, dtype: float64
count    195.000000
mean     160.717949
std       65.918940
min       20.000000
25%      120.000000
50%      160.000000
75%      182.500000
max      450.000000
Name: Height, dtype: float64
guyed lattice
count    76.000000
mean      2.328947
std       1.627990
min       0.000000
25%       1.000000
50%       2.000000
75%       3.250000
max       6.000000
Name: Carriers, dtype: float64
count      76.000000
mean      278.677632
std       258.772161
min        35.000000
25%       147.500000
50%       195.000000
75%       308.250000
max      1339.000000
Name: Height, dtype: float64
utility pole
count    160.000000
mean       0.993750
std        0.079057
min        0.000000
25%        1.000000
50%        1.000000
75%        1.000000
max        1.000000
Name: Carriers, dtype: float6

### Find the closest coordinates within 5 kim radius from the centre(given coordinates) 

In [5]:
#Find the closest coordinates to the given coordinates
def closest_coordinates(lat, long):
    
    coord_diff = (df["Latitude"]-lat).abs() + (df["Longitude"]-long).abs()    #absolute difference between the 2 coordinates
    coord_sorted = coord_diff.argsort()    # sort the coordinates based on closeness to give coordinates
    return df.iloc[coord_sorted[:5]]     #Choose the 10 closest coordinates
    
#Given coordinates
lat = 41.50
long = -72.3
radius_limit = 5000   # 5 km radius from the centre which is the given latitude and longitude
df_closest = closest_coordinates(lat, long)
df_closest

Town   Latitude  Longitude               Tower Type  Height  \
907        Salem  41.502825 -72.297047                 monopole   197.0   
784        Salem  41.487833 -72.313250                 monopole   180.0   
535   Colchester  41.544944 -72.304778                 monopole   180.0   
666  East Haddam  41.471583 -72.323306                 monopole   160.0   
141        Salem  41.468472 -72.273306  self-supporting lattice   190.0   

                carrier_1             carrier_2             carrier_3  \
907       Cingular @ 186'  Cingular/AT&T @ 175'         Sprint @ 195'   
784        Verizon @ 180'                   NaN                   NaN   
535  Cingular/AT&T @ 150'       enertrac @ 169'         Sprint @ 180'   
666           AT&T @ 157'                   NaN                   NaN   
141         Sprint @ 150'       Cingular @ 187'  Cingular/AT&T @ 127'   

           carrier_4       carrier_5 carrier_6 carrier_7  \
907  T-Mobile @ 175'             NaN       NaN       NaN   
784              NaN             NaN       NaN       NaN   
535  T-Mobile @ 138'  Verizon @ 161'       NaN       NaN   
666              NaN             NaN       NaN       NaN   
141  T-Mobile @ 175'             NaN       NaN       NaN   

     New Georeferenced Column  Carriers  
907                         8         4  
784                         8         1  
535                         8         5  
666                         3         1  
141                         8         4

### Let's say for example, T-Mobile Internet Service Provider asks for locations near the given location. The data solution to this question would be the coordinates (41.487, -72.313) and (41.471, -72.323) because the other 3 towers already have T-Mobile as a carrier. <br>
### T-Mobile also wants to know which type of tower they should go for to get maximum results. The data solution to this question would be the tower type "Monopole" as it is the most popular tower type in that area.

In [11]:
#Find the closest coordinates to the given coordinates
def closest_coordinates(lat, long):
    
    coord_diff = (df["Latitude"]-lat).abs() + (df["Longitude"]-long).abs()    #absolute difference between the 2 coordinates
    coord_sorted = coord_diff.argsort()    # sort the coordinates based on closeness to give coordinates
    return df.iloc[coord_sorted[:5]]     #Choose the 10 closest coordinates
    
#Given coordinates
lat = 41.20
long = -72.50
radius_limit = 5000   # 5 km radius from the centre which is the given latitude and longitude
df_closest = closest_coordinates(lat, long)
df_closest

Town   Latitude  Longitude               Tower Type  Height  \
2      Clinton  41.275206 -72.497711  self-supporting lattice   195.0   
4      Clinton  41.278814 -72.525975  self-supporting lattice    67.0   
801  Westbrook  41.290556 -72.468889                 monopole   160.0   
92     Clinton  41.288611 -72.539167  self-supporting lattice   212.0   
844  Westbrook  41.323778 -72.491111                 monopole   160.0   

                   carrier_1      carrier_2        carrier_3       carrier_4  \
2             Verizon @ 162'  Sprint @ 182'  Cingular @ 152'             NaN   
4             T-Mobile @ 60'  Verizon @ 50'              NaN             NaN   
801          T-Mobile @ 145'  Nextel @ 130'       Sprint @ ?  Verizon @ 110'   
92   Verizon @ 208'/130'/60'  Sprint @ 196'  T-Mobile @ 135'   Nextel @ 175'   
844     Cingular/AT&T @ 160'            NaN              NaN             NaN   

           carrier_5 carrier_6 carrier_7  New Georeferenced Column  Carriers  
2                NaN       NaN       NaN                         3         3  
4                NaN       NaN       NaN                         3         2  
801              NaN       NaN       NaN                         3         4  
92   Cingular @ 190'       NaN       NaN                         3         5  
844              NaN       NaN       NaN                         3         1

### Let's say for example, Verizon Internet Service Provider asks for locations near the given location. The data solution to this question would be the coordinates (41.323, -72.491) because the other 4 towers already have Verizon as a carrier. <br>
### Verizon also wants to know which type of tower they should go for to get maximum results. The data solution to this question would be the tower type "Monopole" as it is the only tower without any previous Verizon carrier.